File: algorithm/leetcode16_3sum-closest.cpp
class Solution {
public:
    int threeSumClosest(vector<int>& nums, int target) {
        sort(nums.begin(), nums.end());
        int i, s, e, minDiffer = INT_MAX, record;
        int cursum = nums[0]+nums[1]+nums[2];
        if(target <= cursum)
        	return cursum;
        for(int i = 0; i < nums.size()-2; ++i)
        {
        	s = i+1;
        	e = nums.size()-1;
        	while(s < e)
        	{
        		cursum = nums[i]+nums[s]+nums[e];
        		if(abs(cursum-target) < minDiffer)
        		{
        			minDiffer = abs(cursum-target);
        			record = cursum;
        		}
        		if(cursum < target)
        			++s;
        		else if(cursum == target)
        			return cursum;
        		else
        			--e;
        	}
        }
        return record;
    }
};

File: algorithm/Python3_version/leetcode16.3sum-closest.py
class Solution:
    def threeSumClosest(self, nums: List[int], target: int) -> int:
        nums.sort()
        minDiffer, record = float('inf'), 0
        cursum = nums[0]+nums[1]+nums[2];
        if target <= cursum:
            return cursum
        n = len(nums)
        cursum = nums[n-3]+nums[n-2]+nums[n-1]
        if target >= cursum:
            return cursum
        for i in range(n-2):
            s = i+1
            e = n-1
            while s < e:
                cursum = nums[i]+nums[s]+nums[e]
                if abs(cursum-target) < minDiffer:
                    minDiffer = abs(cursum-target)
                    record = cursum
                if cursum < target:
                    s += 1
                elif cursum == target:
                    return cursum
                else:
                    e -= 1
        return record